### ISSUE TITLE
Spike on deploying multiple jobs in GCP

### ISSUE THREAD

**Issue by mathnathan** [emelex-ai/BRIDGE#190]

**Comment by sharkeyboi** [comment]
I researched the offerings from GCP and found an option we didn't consider which I think might align well with our needs.
https://cloud.google.com/run/docs/create-jobs
Basically Cloud Run has support for running one off Jobs to completion. Out of the box it supports running multiple tasks in parallel for a job. The alternatives I looked at were containerized Compute Engine or Vertex AI, but I think we would end up configuring both to work the same as this does out of the box.
The steps we would need to take are:
1. Containerize the application: Add a Dockerfile with "ENTRYPOINT" set to run our training script. Add Cloud Storage I/O for config loading and saving the model outputs.
2. Create a job in cloud run using this image. Each task within the job would be a training run with a unique configuration.
3. Each task gets assigned a unique index. We will need logic to associate the index with a training configuration (link to cloud storage?)

I'm thinking there's further room for customization, like having separate pretraining and finetuning jobs, with the finetuning being triggered once the pretraining is finished.

**Comment by mathnathan** [comment]
The Cloud Run Jobs do not accept HTTP requests like normal cloud run services. Great, that is fine. However, to schedule a job to run using the Cloud Run Scheduler, you must set the **HTTP method** to POST. Fine, that's how its done, but somewhat inelegant. I'd prefer to just have a yaml file in our repo and run it with a pipeline in github. That is probably possible if we create a service account for our Emelex GitHub organization? Given the infrequent nature of the jobs, we can also just clickops the job launch from the UI. I'm curious about costs. I just threw a few numbers into the GCP pricing calculator to see how much it would cost. @sharkeyboi, can you do the same I'd like to compare notes, but I don't want to bias you based on what I put in. What do you think?

Big picture though, given the price is reasonable, this seems like a solid solution!

**Comment by sharkeyboi** [comment]
Hey @mathnathan, I ran the pricing numbers assuming 800,000 CPU seconds for the full training pipeline (10 seconds per epoch for 50 pretraining runs and 1100 fine tuning runs), the cpu + memory cost is about $10.7, based on their price of $0.00001800 / vCPU-second and $0.00000200 / GiB-second.

In terms of launching the jobs, my plan was just to start the jobs via a shell/gcloud script. I don't think we would need to use Cloud Scheduler as we don't need these jobs to run on a cron job schedule, nor do I think we would need a github pipeline as we wouldn't need to automate running the jobs when pushing a new commit. Unless you are envisioning we need to be running this training in a more scheduled way than I'm envisioning.

**Comment by mathnathan** [comment]
Great, I think my estimates were a bit more conservative than yours, which is good. It is cheap and and doesn't even include the free-tier, e.g. (First 240,000 vCPU-seconds free per month, and First 450,000 GiB-seconds free per month). Should be a great way to get quick results. I'll update later today with what I'm able to finish. Thanks for digging into this for us!
